src/events/message/commands/fun/rankup.ts

```diff
@@ -0,0 +1,184 @@
+import { z } from "zod";
+import fetch from "node-fetch";
+import { Message, CommandArg, word } from "../types";
+
+const USAGE: string =
+  "Usage: `?rankup [<username> --target=(username|rank) --language=(py|js|...) --mode={least|each|spread} --limit=(1-8[kyu])]`";
+
+/*
+`?rankup` calculates the number of katas of each rank required for a user to rank up.
+This command is just for fun, however should increase competition between users,
+and (hopefully) increase motivation.
+
+If no username is passed, server nickname of user is taken instead.
+
+Options:
+  <username>   The username to check. Defaults to users discord server nickname
+  --target=    The rank to reach, or a username of a user to overtake. Defaults to users next rank
+  --language=  The language to inspect. Defaults to "overall"
+  --mode=      Calculation mode.
+                  Least favours 1kyus for least overall katas.
+                  Each finds number of katas per rank individually
+                  Spread (default) prioritises high katas, but also distributes points to lower ones
+  --limit=     The max rank to use in calculations ([1-8]kyu)
+*/
+
+enum Mode {
+  Least = "least",
+  Each = "each",
+  Spread = "spread", // default
+}
```

Comment:
I usually avoid `enum` because the equivalent doesn't exist in JavaScript, and I forget how it's represented there.

I usually do:
```typescript
const LEAST = "least";
const EACH = "each";
const SPREAD = "spread";
type Mode = LEAST | EACH | SPREAD;
```
And define a [type predicate](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards):
```typescript
function isMode(value: unknown): value is Mode {
  return typeof value == "string" && (value == LEAST || value == EACH || value == SPREAD);
  // Or define `MODES` and do `MODES.indludes(value)`
}
```
This will make sure the value is `Mode`.